crates/sui-types/src/temporary_store.rs

```diff
@@ -689,6 +689,150 @@ impl<S> TemporaryStore<S> {
+                            // Note: this assumes that the only immutable objects an epoch change tx can update are system packages,
+                            // but in principle we could allow others.
+                            assert!(
+                                is_system_package(*id),
+                                "Only system packages can be upgraded"
+                            );
+                        }
+                    }
+                }
+                WriteKind::Create | WriteKind::Unwrap => {
+                    // created and unwrapped objects were not inputs to the tx
+                    // or dynamic fields accessed at runtime, no ownership checks needed
+                }
+            }
+        }
+
+        for (id, (_version, kind)) in &self.deleted {
+            if authenticated_objs.contains(id) || gas_objs.contains(id) {
+                continue;
+            }
+            match kind {
+                DeleteKind::Normal | DeleteKind::Wrap => {
+                    // get owner at beginning of tx
+                    let old_obj = self.store.get_object(id)?.unwrap();
+                    match &old_obj.owner {
+                        Owner::ObjectOwner(_) => {
+                            objs_to_authenticate.push(*id);
+                        }
+                        Owner::AddressOwner(_) | Owner::Shared { .. } => {
+                            unreachable!("Should already be in authenticated_objs")
+                        }
+                        Owner::Immutable => unreachable!("Immutable objects cannot be deleted"),
+                    }
+                }
+                DeleteKind::UnwrapThenDelete => {
+                    // unwrapped-then-deleted object was not an input to the tx,
+                    // no ownership checks needed
+                }
+            }
+        }
+        Ok((objs_to_authenticate, authenticated_objs))
+    }
+
+    // check that every object read is owned directly or indirectly by sender, sponsor, or a shared object input
+    pub fn check_ownership_invariants(
+        &self,
+        sender: &SuiAddress,
+        gas: &[ObjectRef],
+        is_epoch_change: bool,
+    ) -> SuiResult<()> {
+        let (mut objects_to_authenticate, mut authenticated_objects) =
+            self.get_objects_to_authenticate(sender, gas, is_epoch_change)?;
+
+        // Map from an ObjectID to the ObjectID that covers it.
+        let mut covered = BTreeMap::new();
+        // iterate until we we have authenticated all objects, or reached a fixed point with remaining objects to authenticate
+        while let Some(to_authenticate) = objects_to_authenticate.pop() {
+            let old_obj = self.store.get_object(&to_authenticate)?.expect(
+                "Object authentication failure: could not find parent object {id:?},\n\
```

Comment:
How does this work for things like `Tables`? The "parent" will be a `UID` that doesn't actually appear in storage.
How are more tests not failing these checks?

Reply:
Would also appreciate your advice on how to distinguish this case from an authentication bug.

Reply:
I... do not know. The most expensive option is to search for all UIDs contained in input objects and loaded objects. And I am not sure how to do something more efficient

Reply:
> Do we have any tests using `Table` that run the full executor pipeline?

We might not. Well... it might be the simtest that is breaking, since I think the system state object now uses `Bag`. 
A high-pri sort of test auditing should be to get the sui-adapter-transactional-tests onto the authority pipeline

Reply:
> Would also appreciate your advice on how to distinguish this case from an authentication bug.

Maybe just assume things are okay if the object isn't in storage? 
Without actually visiting all of the object fields (recursively), I do not think it is possible to differentiate between a dynamic field access on a wrapped object, and an authentication bug where the parent has been wrapped.

Reply:
> Maybe just assume things are okay if the object isn't in storage?

Good call--this seems like the path of least resistance. I have an idea for a different, more lightweight dynamic check to look at that.